### PR TITLE
Supporting Algo Management API's

### PR DESCRIPTION
**Pull request by besirkurtulmus**
The new methods are basically wrapper around the generated python client, that can be found [here](https://github.com/algorithmiaio/algorithmia-api-client/tree/master/python).

I've added the `.create()`, `.update()`, `.publish()`, `.info()`, `.versions()` & `.compile()` methods under `client.algo()`.

### CODE REVIEW

File: Algorithmia/algorithm.py [hunk 1]
```diff
@@ -17,6 +21,10 @@ def __init__(self, client, algoRef):
         if m is not None:
             self.client = client
             self.path = m.group(1)
+            self.username = self.path.split("/")[0]
```
Comment (zeryx) [inline Algorithmia/algorithm.py]: Is it safe to assume here that the algorithm URI structure will always remain the same? Is there a world where this is brittle(supporting unicode algorithm names/usernames?)

File: Algorithmia/algorithm.py [hunk 2]
```diff
@@ -29,6 +37,97 @@ def set_options(self, timeout=300, stdout=False, output=OutputType.default, **qu
...
+            error_message = json.loads(e.body)["error"]["message"]
+            raise ApiError(error_message)
+
+    # Update the settings in an algorithm
+    def update(self, details={}, settings={}, version_info={}):
+        detailsObj = Details(**details)
+        settingsObj = Settings(**settings)
+        createRequestVersionInfoObj = CreateRequestVersionInfo(**version_info)
+        update_parameters = {"details": detailsObj, "settings": settingsObj, "version_info": createRequestVersionInfoObj}
+        update_request = UpdateRequest(**update_parameters)
```
Comment (zeryx) [inline Algorithmia/algorithm.py]: Small nit:
Some of your variables are camel case, and some are snake case - I think most of the time you snake case for dicts, but not always. Might be useful to be consistent here.

I think the python default is always snake case?

File: Algorithmia/algorithm.py (hunk 1 above)
Comment (besirkurtulmus) [reply Algorithmia/algorithm.py]: I would assume that if the URI structure ever changes, line 19 would also break (which has been part of the client forever now).

Other than that point, I don't think we'll ever change it.

File: Algorithmia/algorithm.py (hunk 2 above)
Comment (besirkurtulmus) [reply Algorithmia/algorithm.py]: Good point. Also as far as I can tell, the client has used both :/

Look at line 19-30. I could change the style for the code I wrote. I personally prefer using underscores. It makes reading the code easier.

File: Algorithmia/algorithm.py [hunk 3]
```diff
@@ -29,6 +37,97 @@ def set_options(self, timeout=300, stdout=False, output=OutputType.default, **qu
         self.query_parameters.update(query_parameters)
         return self
 
+    # Create a new algorithm
+    def create(self, details={}, settings={}, version_info={}):
+        detailsObj = Details(**details)
+        settingsObj = SettingsMandatory(**settings)
+        createRequestVersionInfoObj = CreateRequestVersionInfo(**version_info)
+        create_parameters = {"name": self.algoname, "details": detailsObj, "settings": settingsObj, "version_info": createRequestVersionInfoObj}
```
Comment (kennydaniel) [inline Algorithmia/algorithm.py]: Where does it get the username from? How does this work if I'm creating an org-owned algo?

Comment (kennydaniel) [reply Algorithmia/algorithm.py]: Nvm, now I see it below.

File: Algorithmia/algorithm.py [hunk 4]
```diff
@@ -29,6 +37,97 @@ def set_options(self, timeout=300, stdout=False, output=OutputType.default, **qu
...
+        try:
+            # Get Algorithm
+            if algo_hash:
+                api_response = self.client.manageApi.get_algorithm_hash_version(self.username, self.algoname, algo_hash)
+            else:
+                api_response = self.client.manageApi.get_algorithm(self.username, self.algoname)
+            return api_response
+        except ApiException as e:
+            error_message = json.loads(e.body)["error"]["message"]
+            raise ApiError(error_message)
```
Comment (kennydaniel) [inline Algorithmia/algorithm.py]: The way we're handling exceptions is throwing away information. What about the error code? stacktrace? etc? Probably doesn't belong in this PR, but something to think about.

File: Algorithmia/algorithm.py (hunk 2 above)
Comment (neildf) [reply Algorithmia/algorithm.py]: Per python style guide variables should be snake case:

https://www.python.org/dev/peps/pep-0008/#function-and-variable-names
